lib/promiscuous/kafka/poseidon.rb

```diff
@@ -0,0 +1,149 @@
+                                          :ack_timeout_ms => 1000,
+                                          :socket_timeout_ms => Promiscuous::Config.socket_timeout)
+    connection
+  end
+
+  def connect
+    @connection = new_connection
+    # Poseidon.logger = Logger.new(STDOUT).tap { |l| l.level = 0 }
+  end
+
+  def disconnect
+    @connection_lock.synchronize do
+      return unless connected?
+      @connection.shutdown
+      @connection = nil
+    end
+  end
+
+  #TODO: broken
+  def connected?
+    !!@connection.try(:connected?)
+  end
+
+  def raw_publish(options)
+    ok = false
+    10.times do
+      ok = @connection.send_messages([Poseidon::MessageToSend.new(options[:topic], options[:payload], options[:key])])
+      break if ok
+      sleep(1)
+    end
+    raise "Unable to send messages" if !ok
+  end
+
+  def publish(options={})
+    @connection_lock.synchronize do
+      raw_publish(options)
+      options[:on_confirm].call if options[:on_confirm]
+      Promiscuous.debug "[publish] [kafka] #{options[:topic]}/#{options[:topic_key]} #{options[:payload]}"
+    end
+  rescue Exception => e
+    Promiscuous.warn("[publish] Failure publishing to kafka #{e}\n#{e.backtrace.join("\n")}")
+    e = Promiscuous::Error::Publisher.new(e, :payload => options[:payload])
+
+    if options[:async]
+      Promiscuous::Config.error_notifier.call(e)
+    else
+      raise e
+    end
+  end
+
+  module Subscriber
+    def subscribe(topic)
+      @consumer = ::Poseidon::ConsumerGroup.new(Promiscuous::Config.app,
+                                                Promiscuous::Config.kafka_hosts,
+                                                Promiscuous::Config.zookeeper_hosts,
+                                                topic, :trail => Promiscuous::Config.test_mode, :max_wait_ms => 10)
+    end
+
+    def fetch_and_process_messages(&block)
+      Promiscuous.debug "[kafka] Fetching messages topic:#{@consumer.topic} #{Thread.current}"
+      @consumer.fetch(:commit => false) do |partition, payloads|
+        Promiscuous.debug "[kafka] Received #{payloads.count} payloads topic:#{@consumer.topic} #{Thread.current}"
+        payloads.each do |payload|
+          Promiscuous.debug "[kafka] Fetched '#{payload.value}' topic:#{@consumer.topic} offset:#{payload.offset} parition:#{partition}"
+          block.call(MetaData.new(@consumer, partition, payload.offset), payload)
+        end
+      end
+    end
+
+    def delete_queues
+      # TODO: implement, possibly keep inside of the specs
+      # Promiscuous::Config.subscriber_topics.each { |queue| queue.try(:delete) }
+    end
+
+    class MetaData
+      def initialize(consumer, partition, offset)
+        @consumer = consumer
+        @partition = partition
+        @offset = offset
+
+        Promiscuous.debug "[kafka] [metadata] topic:#{@consumer.topic} offset:#{offset} partition:#{partition}"
+      end
+
+      # TODO: rename to advance_offset or commit?
+      def ack
```

Comment:
It has to remain this way if we're going to have pluggable backends with the same interface